### PR TITLE
gh-88239: Use sqlite3_stmt_busy() to determine if statements are in use

### LINKED ISSUE #88239: [sqlite3] drop statement in_use field in favour of sqlite3_stmt_busy()
BPO | [44073](https://bugs.python.org/issue44073)
--- | :---
Nosy | @berkerpeksag, @serhiy-storchaka, @erlend-aasland
PRs | <li>python/cpython#25984</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/erlend-aasland'
closed_at = None
created_at = <Date 2021-05-08.07:47:59.789>
labels = ['extension-modules', 'easy', 'type-feature', '3.11']
title = '[sqlite3] drop statement in_use field in favour of sqlite3_stmt_busy()'
updated_at = <Date 2021-08-19.20:10:44.154>
user = 'https://github.com/erlend-aasland'
```

bugs.python.org fields:
```python
activity = <Date 2021-08-19.20:10:44.154>
actor = 'erlendaasland'
assignee = 'erlendaasland'
closed = False
closed_date = None
closer = None
components = ['Extension Modules']
creation = <Date 2021-05-08.07:47:59.789>
creator = 'erlendaasland'
dependencies = []
files = []
hgrepos = []
issue_num = 44073
keywords = ['patch', 'easy (C)']
message_count = 3.0
messages = ['393241', '393243', '399932']
nosy_count = 3.0
nosy_names = ['berker.peksag', 'serhiy.storchaka', 'erlendaasland']
pr_nums = ['25984']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue44073'
versions = ['Python 3.11']
```

</p></details>

### PR DESCRIPTION
**Pull request by erlend-aasland**
Fixes #88239 

* Issue: gh-88239

### CODE REVIEW

**Comment by erlend-aasland** [comment]
No user-visible change, so I'm adding the _skip news_ label.

Marking as draft until I've verified all the exit paths in _pysqlite_query_execute().

**Comment by erlend-aasland** [comment]
cc. @corona10

**Comment by erlend-aasland** [comment]
> Marking as draft until I've verified all the exit paths in _pysqlite_query_execute().

I've left the exit paths as they are to avoid changing current behaviour. We can rework `_pysqlite_query_execute` later. There's much to do there :)

**Comment by erlend-aasland** [comment]
@pablogsal, are you comfortable with reviewing this, wrt. SQLite API? The change is very straight-forwared, so it should be easy to review.

Here's a link to the relevant API: https://sqlite.org/c3ref/stmt_busy.html

**Review by pablogsal — APPROVED** [review]

**Comment by pablogsal** [comment]
Do we have tests covering this behaviour from before?

**Comment by erlend-aasland** [comment]
> Do we have tests covering this behaviour from before? 

The if statement in query execute lacks test coverage. I've got one lying around from the sqlite3 coverage issue. I'll add it to the PR once I'm back at my computer. Thanks!

**Comment by pablogsal** [comment]
I would like to ame sure that we have coverage for this to ensure that we don't introduce regressions (although we shouldn't if I read this PR correctly).

File: Lib/sqlite3/test/dbapi.py
Comment (erlend-aasland) [inline Lib/sqlite3/test/dbapi.py]: This test exercises this branch:
https://github.com/python/cpython/blob/006fd869e4798b68e266f5de89c83ddb531a756b/Modules/_sqlite/cursor.c#L536-L542

**Comment by erlend-aasland** [comment]
> I would like to ame sure that we have coverage for this to ensure that we don't introduce regressions (although we shouldn't if I read this PR correctly).

Of course.

BTW, I restored the optimisation of `pysqlite_statement_reset`: only call `sqlite3_reset` if needed. Calling it every time would acquire and release the SQLite mutex.

File: Modules/_sqlite/cursor.c
Comment (erlend-aasland) [inline Modules/_sqlite/cursor.c:536]: `sqlite3_stmt_busy` also does a NULL pointer check, but we'll avoid a function call if we do it ourselves. Resetting a statement is a frequent operation.

File: Modules/_sqlite/statement.c
Comment (erlend-aasland) [inline Modules/_sqlite/statement.c:377]: `sqlite3_stmt_busy` also does a NULL pointer check, but we'll avoid a function call if we do it ourselves. Resetting a statement is a frequent operation.

Comment (erlend-aasland) [reply Modules/_sqlite/statement.c:377]: Aah, if I'm to add this optimisation back, I need to harden the exit points in the query execute code. But that's a nice side effect. I'll mark this as draft until I'm done.

Comment (erlend-aasland) [reply Modules/_sqlite/statement.c]: Actually, I think it's better to leave the optimisation out and rather optimise the API usage in a later PR; that is, only call reset when needed.

File: Modules/_sqlite/statement.c
Comment (erlend-aasland) [inline Modules/_sqlite/statement.c:377]: This may lead to `sqlite3_reset` being called multiple times on a statement, but there is ongoing effort to optimise this part of the SQLite API usage ([issue 43350](https://bugs.python.org/issue43350)).

**Comment by erlend-aasland** [comment]
Putting this on hold until we've cleaned up `sqlite3_reset` usage.

**Comment by erlend-aasland** [comment]
On hold until #27844 is merged.

**Comment by erlend-aasland** [comment]
Buildbots are happy. I'll be merging this to `main` in a couple of days. This will not be backported.

**Pull request merged**
